massa-bootstrap/src/server.rs

```diff
@@ -14,363 +42,470 @@ use massa_signature::KeyPair;
     consensus_controller: Box<dyn ConsensusController>,
     network_command_sender: NetworkCommandSender,
     final_state: Arc<RwLock<FinalState>>,
-    establisher: Establisher,
-    manager_rx: mpsc::Receiver<()>,
-    bind: SocketAddr,
+    listener_rx: Receiver<BsConn>,
+    stopper_rx: Receiver<()>,
+    allow_block_list: SharedAllowBlockList<'a>,
     keypair: KeyPair,
     bootstrap_config: BootstrapConfig,
     version: Version,
     ip_hist_map: HashMap<IpAddr, Instant>,
+    runtime: Runtime,
 }
 
-#[allow(clippy::result_large_err)]
-#[allow(clippy::type_complexity)]
-fn reload_whitelist_blacklist(
-    whitelist_path: &PathBuf,
-    blacklist_path: &PathBuf,
-) -> Result<(Option<HashSet<IpAddr>>, Option<HashSet<IpAddr>>), BootstrapError> {
-    let whitelist = if let Ok(whitelist) = std::fs::read_to_string(whitelist_path) {
-        Some(
-            serde_json::from_str::<HashSet<IpAddr>>(whitelist.as_str())
-                .map_err(|_| {
-                    BootstrapError::GeneralError(String::from(
-                        "Failed to parse bootstrap whitelist",
-                    ))
-                })?
-                .into_iter()
-                .map(normalize_ip)
-                .collect(),
-        )
-    } else {
-        None
-    };
+impl BootstrapServer<'_> {
+    async fn run_updater(
+        mut list: SharedAllowBlockList<'_>,
+        interval: Duration,
+    ) -> Result<(), String> {
+        let mut interval = tokio::time::interval(interval);
+        loop {
+            interval.tick().await;
+            // TODO: loop interval here is tick + update time. Implement a state-based,
+            // rather than time-based, trigger (such as delta-count);
+            list.update()?;
+        }
+    }
 
-    let blacklist = if let Ok(blacklist) = std::fs::read_to_string(blacklist_path) {
-        Some(
-            serde_json::from_str::<HashSet<IpAddr>>(blacklist.as_str())
-                .map_err(|_| {
-                    BootstrapError::GeneralError(String::from(
-                        "Failed to parse bootstrap blacklist",
-                    ))
-                })?
-                .into_iter()
-                .map(normalize_ip)
-                .collect(),
-        )
-    } else {
-        None
-    };
-    Ok((whitelist, blacklist))
-}
+    /// Listens on a channel for incoming connections, and loads them onto a crossbeam channel
+    /// for the main-loop to process.
+    ///
+    /// Ok(Ok(())) listener closed without issue
+    /// Ok(Err((dplx, address))) listener accepted a connection then tried sending on a disconnected channel
+    /// Err(..) Error accepting a connection
+    async fn run_listener(
+        mut listener: Listener,
+        listener_tx: Sender<BsConn>,
+    ) -> Result<Result<(), BsConn>, Box<BootstrapError>> {
+        loop {
+            let msg = listener.accept().await.map_err(BootstrapError::IoError)?;
+            match listener_tx.send(msg) {
```

Comment:
Why does the listener send back the connection to the main loop instead of using a shared data to check if he can accept it and spawn the thread for the connection himself? It will avoid this communication

Reply:
:thinking:

Reply:
I don't understand your emoji 😆 . It's related to https://github.com/massalabs/massa/pull/3554#discussion_r1114320260

Reply:
If the `.await` wasn't needed, then we could put the `.accept()` at the top of the main-loop. That could work well under the assumption that the time taken to sanity-check the client before the bootstrap thread is spawned is sufficiently short.

...and if we are not happy to make that assumption, we could move a whole bunch of it into the handler itself (though that might mean having to set-up something to handle the join() results on that. We should be doing that anyway, but might be a bit premature for this PR)

If we are happy to block on this accept inside the main-loop, we could do something like this?

```rust
let acceptance = Runtime::new()?.block_on(/*loop-body here */);
```

Reply:
> I don't understand your emoji laughing . It's related to [#3554 (comment)](https://github.com/massalabs/massa/pull/3554#discussion_r1114320260)

it's the thinking one. Intended to communicate that "...huh" moment that gives food for thought.

I've since made a `close_and_send_error` helper that

- consumes the server binder
- spawns a thread, and makes an async runtime in it
- sends the message
- runs a passed-in closure (which is typically to run a logging macro)

Can we resolve this thread, and move discussion to https://github.com/massalabs/massa/pull/3554#discussion_r1114320260 ?, (make a reference to this thread over there)